frida_tools/tracer.py

```diff
@@ -581,6 +626,13 @@ def _create_trace_script(self):
 var pending = [];
 var timer = null;
 
+// Create namespaces for user functions ('F') and data ('D'), pre-populate if relevant
+state.F = {}
+state.D = {}
+state.D.NEWLINE = String.fromCharCode(10);
```

Comment:
How about naming these `$userFunctions` and `$userData`, assuming they're "special"? (In any case we shouldn't violate the style by using PascalCase for properties – properties should be camelCase.) Are these meant to be accessed by the user or are they internals?

Reply:
The are meant to be access by the user.  So, for example, a developer might create a function called `dump_function_parameters` (a real example we have done).  The init-session JS script defines this function, storing it in `state`.  I intentionally wanted to create the shortest possible name (`state.F`) to reduce noise in handler functions that call this function.  The function call:

`state.F.dump_function_parameters (parms)`

is shorter than:

`state.userFunctions.dump_function_parameters (parms)`.

Having said that, I actually like your suggestion, for clarity's sake.  If you have no problem with the possible verbosity, I think `userFunctions` and `userData` are quite acceptable.  Awaiting your decision.

Reply:
Ah I see. Thanks for clarifying! Thinking about it more, we could also just evaluate those scripts in the global context (which I guess is already happening?), and have them be globally visible. That way any variables and functions are globally visible and conveniently accessible from the hooks. Would that work?

Reply:
I wanted to provide the developer with a convenient namespace for functions (`'state.F'`) and data (`'state.D'`).  This would prevent 'polluting' the global namespace.

To clarify this and other issues, I provide one of the `init-session` JS files I use in all my frida-trace work.  I now typically invoke frida-trace with:

`frida-trace -init-session /<path>/frida-library.general.js ...`.

In this way, whenever a handler's `onEnter` / `onLeave` is entered, I know that I can invoke any of the functions:

```
state.F.general_clone_args()
state.F.general_hexdump_lines(outlines, desc, nptr, len, indent_string)
state.F.general_hexdump(log, desc, nptr, len)
state.F.general_decode_bitflags(value, dict)
state.F.general_decode_enum(value, dict)
state.F.general_backtrace(log, desc, my_this)
state.F.process_function_parameters(outlines, funcname, my_args, parmdata, flags)
state.F.dump_function_parameters(log, funcname, my_args, parmdata, flags)
```

Here's the `init-session` file I referred to above (excuse the extra whitespace <g>).  I typically specify another `init-session` file on the command line for non-general functions (e.g., functions specific to a particular SO / DLL I'm working on):

```
// Define useful general-purpose Frida handler functions

//--------------------------------------------------------------------
// general_clone_args()
//--------------------------------------------------------------------

state.F.general_clone_args = function (args, num_args, my_this)
{
    if (my_this.args === undefined)
    {
        my_this.args = []

        for (var i=0; i<num_args; ++i)
        {
            my_this.args.push (args[i]);
        }
    }
};

//--------------------------------------------------------------------
// general_return_hexdump_lines()
//--------------------------------------------------------------------

state.F.general_hexdump_lines = function (outlines, desc, nptr, len, indent_string)
{
    if (nptr === undefined)
    {
        outlines.push (indent_string + 'INFO: (general_hexdump: ' + desc + ') nptr is UNDEFINED');
    }
    else if (nptr.isNull() === true)
    {
        outlines.push (indent_string + 'INFO: (general_hexdump: ' + desc + ') nptr is NULL');
    }
    else
    {
        outlines.push (desc);
        var s = hexdump (nptr, {length: len});
        outlines.push (indent_string + s);
    }
};

//--------------------------------------------------------------------
// general_hexdump()
//--------------------------------------------------------------------

state.F.general_hexdump = function (log, desc, nptr, len)
{
    var outlines = [];

    state.F.general_hexdump_lines (outlines, desc, nptr, len, '');
    log (outlines.join(NEWLINE));
};

//--------------------------------------------------------------------
// general_decode_bitflags()
//--------------------------------------------------------------------

state.F.general_decode_bitflags = function (value, dict)
{
    var s = '';

    for (var key in dict)
    {
        if (value & key)
        {
            if (s.length > 0)
            {
                s = s + ' | ';
            }

            s = s + dict[key];
        }
    }

    return s;
};

//--------------------------------------------------------------------
// general_decode_enum()
//--------------------------------------------------------------------

state.F.general_decode_enum = function (value, dict)
{
    for (var key in dict)
    {
        if (value == key)
        {
            return dict[key];
        }
    }

    return '<UNKNOWN>';
};

//--------------------------------------------------------------------
// general_backtrace()
//--------------------------------------------------------------------

state.F.general_backtrace_lines = function (outlines, desc, my_this)
{
    if (my_this !== undefined)
    {
        if (my_this.context !== undefined)
        {
            outlines.push (NEWLINE + '[*] backtrace (' + desc + '): ' + Thread.backtrace (my_this.context, Backtracer.ACCURATE).map(DebugSymbol.fromAddress).join(NEWLINE) + NEWLINE);
        }
        else
        {
            outlines.push (NEWLINE + '[*] backtrace (' + desc + '): No context found' + String.fromCharCode (10));
        }
    }
};

state.F.general_backtrace = function (log, desc, my_this)
{
    var outlines = [];
    state.F.general_backtrace_lines (outlines, desc, my_this);
    log (outlines.join (NEWLINE));
};

//--------------------------------------------------------------------
// process_function_parameters()
//
// This function accepts a function's name and its parameter metadata,
// performing the requested <flags> and writing any output to <log>.
//
// The format of <parmdata> is an array of objects, one object for each
// function parameter.
//
// Here is an example of invoking this function: <parmdata> example
//
//     process_function_parms (log,
//                             'ff_get_buffer',
//                             [
//                                 {name : 'avctx', type : 'AVCodecContext *'},
//                                 {name : 'frame', type : 'AVFrame *'},
//                                 {name : 'flags', type : 'int'}
//                             ],
//                             HEX_DUMP);
//
// Note:
//
//     (1) By default all parameter types are assumed to be 32-bit
//         (4 bytes).  We will enhance this function to support 64-bit
//         (8 byte) when the need arises.
//
//     (2) The number of asterisks in the parameter type string dictates
//         whether it is a non-ptr, a simple pointer, a pointer-to-a-pointer,
//         etc.  Thus, 'int' is a non-ptr, 'AVFrame *' is a simple pointer,
//         and 'undefined4 **' is a pointer-to-a-pointer.
//
//--------------------------------------------------------------------

state.D.GENERAL_HEX_DUMP     = 0x01;
state.D.GENERAL_BRIEF_OUTPUT = 0x02;

state.F.process_function_parameters = function (outlines, funcname, my_args, parmdata, flags)
{
    //----------------------
    //----------------------
    function ComputeNumberDereferencesInType (type)
    {
        var i;
        var count = 0;

        for (i=0; i<type.length; ++i)
        {
            if (type[i] == '*')
            {
                ++count;
            }
        }

        return count;
    }

    //----------------------
    //----------------------
    function process_data (outlines, native_pointer, pos, data, isBriefMode)
    {
        var i;
        var s = '';
        s = s + 'parameter ' + pos + ':';

        if ('type' in data)
        {
            s = s + ' ' + data.type;
        }

        if ('name' in data)
        {
            s = s + ' ' + data.name;
        }

        if (!isBriefMode)
        {
            // Caller did not specify brief output, add newline for readability
            s = NEWLINE + s;
        }
        outlines.push (s);

        var numDerefs = ComputeNumberDereferencesInType (data.type);
        var p = ptr(native_pointer);

        if (numDerefs === 0)
        {
            // Different output for non-pointer
            outlines.push ('\t[0] ' + p.toString());
            return;
        }

        for (i=0; i<numDerefs; ++i)
        {
            outlines.push ('\t[' + i + '] ' + p.toString());

            if (p.isNull() === true)
                break;

            if (i < numDerefs - 1)
                p = p.readPointer();
        }

        if (p.isNull() === false)
        {
            var t = '\t[' + numDerefs + '] ';
            var len;

            if ('len' in data)
                len = data.len;
            else
                len = 32;
            state.F.general_hexdump_lines (outlines, t, p, len, '\t\t');
        }
    }

    // Main portion of function
    var isBriefMode = (flags & state.D.GENERAL_BRIEF_OUTPUT);

    outlines.push (NEWLINE);
    outlines.push ('==========  BEGIN processing ' + funcname + ' (' + parmdata.length + ' parameters)' + '  ==========================');

    for (var i=0; i<parmdata.length; ++i)
    {
        var data = parmdata[i];
        process_data (outlines, my_args[i], i, data, isBriefMode);
    }

    var t = '==========  END   processing parameters: ' + funcname + '  ==========================' + String.fromCharCode (10);
    if (!isBriefMode)
        t = NEWLINE + t;
    outlines.push (t);
};

//--------------------------------------------------------------------
// dump_function_parameters()
//--------------------------------------------------------------------

state.F.dump_function_parameters = function (log, funcname, my_args, parmdata, flags)
{
    // Optional values
    flags = flags || 0;

    var outlines = [];
    state.F.process_function_parameters (outlines, funcname, my_args, parmdata, state.D.GENERAL_HEX_DUMP | flags);
    log (outlines.join (NEWLINE));
};
```

Note the use of `state.D` to store global constants used by the `state.F.dump_function_parameters` function.

So these functions **are** globally visible and accessible, but tucked under `state.F`.  Of course, the developer could decide to place `general_clone_args()` directly under `state` rather than `state.F`.  Do we want to 'help' the developer by creating `state.F` or to leave it up to him where he wants to place his functions?

Reply:
I think we want to leave it up to the developer, and make it very easy to add globally visible things, as frida-trace scripts are a bit "quick and dirty" in nature. :-)

By the way, as a side-note, function names and variable names should be camelCase in idiomatic JavaScript code. (Obviously users can do what they want but we should be careful so we use this style consistently in our code, to encourage following the most common style.)

Reply:
You are right, the user should be free to store function and data object wherever he wants to.  I have therefore remove references to `state.F` and `state.D` in `tracer.py`.  Should he want to use a dedicated folder under `state`, he can define the folder(s) under state in an `--init-session` on the command line -- in this way, all handlers can assume the folders exist and will not need to test for their existence.

Reply:
Sounds good!

Reply:
Point well made and taken: if and when I get to document examples of init-session JS files, the examples should adhere to the project's JavaScript coding standards.

Reply:
🙌 Just to clarify, the camelCase style is quite pervasive across JavaScript ecosystems and projects, both web and Node.js – it's not something I'm advocating because I'm a fan of that particular style. :-) (If designing a language from scratch I would probably pick snake_case for variables and function/method names.)